Classes/CatchUpHook/DocumentUriPathProjectionHook.php

```diff
@@ -0,0 +1,263 @@
+            default => null
+        };
+
+    }
+
+    public function onAfterEvent(EventInterface $eventInstance, EventEnvelope $eventEnvelope): void
+    {
+        match ($eventInstance::class) {
+            NodeAggregateWasRemoved::class => $this->onAfterNodeAggregateWasRemoved($eventInstance),
+            NodePropertiesWereSet::class => $this->onAfterNodePropertiesWereSet($eventInstance),
+            NodeAggregateWasMoved::class => $this->onAfterNodeAggregateWasMoved($eventInstance),
+            default => null
+        };
+    }
+
+    public function onBeforeBatchCompleted(): void
+    {
+        // Nothing to do here
+    }
+
+    public function onAfterCatchUp(): void
+    {
+        // Nothing to do here
+    }
+
+    private function onBeforeNodeAggregateWasRemoved(NodeAggregateWasRemoved $event): void
+    {
+        if (!$this->isLiveContentStream($event->contentStreamId)) {
+            return;
+        }
+
+        $this->documentNodeInfosBeforeRemoval = [];
+
+        foreach ($event->affectedCoveredDimensionSpacePoints as $dimensionSpacePoint) {
+            $node = $this->tryGetNode(fn() => $this->getState()->getByIdAndDimensionSpacePointHash(
+                $event->nodeAggregateId,
+                $dimensionSpacePoint->hash
+            ));
+            if ($node === null) {
+                // Probably not a document node
+                continue;
+            }
+
+            $this->nodeRedirectService->appendAffectedNode(
+                $node,
+                $this->getNodeAddress($event->contentStreamId, $dimensionSpacePoint, $node->getNodeAggregateId()),
+                $this->getContentRepositoryId()
+            );
+            $this->documentNodeInfosBeforeRemoval[$dimensionSpacePoint->hash][] = $node;
+
+            $descendantsOfNode = $this->getState()->getDescendantsOfNode($node);
+            array_map(
+                function ($descendantOfNode) use ($event, $dimensionSpacePoint) {
+                    $this->nodeRedirectService->appendAffectedNode(
+                        $descendantOfNode,
+                        $this->getNodeAddress($event->contentStreamId, $dimensionSpacePoint, $descendantOfNode->getNodeAggregateId()),
+                        $this->getContentRepositoryId()
+                    );
+                    $this->documentNodeInfosBeforeRemoval[$dimensionSpacePoint->hash][] = $descendantOfNode;
+                },
+                iterator_to_array($descendantsOfNode));
+        }
+    }
+
+    private function onAfterNodeAggregateWasRemoved(NodeAggregateWasRemoved $event): void
+    {
+        if (!$this->isLiveContentStream($event->contentStreamId)) {
+            return;
+        }
+
+        foreach ($event->affectedCoveredDimensionSpacePoints as $dimensionSpacePoint) {
+            $documentNodeInfosBeforeRemoval = $this->documentNodeInfosBeforeRemoval[$dimensionSpacePoint->hash];
```

Comment:
```suggestion
            if (!array_key_exists($dimensionSpacePoint->hash, $this->documentNodeInfosBeforeRemoval)) {
                continue;
            }
            $documentNodeInfosBeforeRemoval = $this->documentNodeInfosBeforeRemoval[$dimensionSpacePoint->hash];
```

Without this check, a lot of behat tests fail for me with

> Warning: Undefined array key "[...]" in DocumentUriPathProjectionHook.php line 122